File: java-workshop/part-3-mocks-clean/src/main/java/domain/Order.java
package domain;

import java.util.List;
public record Order(
    String id,
    String customerId,
    List<OrderItem> items,
    OrderStatus status,
    double totalPrice,
    String transactionId,
    String trackingNumber
) {
    public Order {
        items = List.copyOf(items);
    }

    public Order confirm(double totalPrice) {
        return new Order(id, customerId, items, OrderStatus.CONFIRMED, totalPrice, transactionId, trackingNumber);
    }

    public Order reject() {
        return new Order(id, customerId, items, OrderStatus.REJECTED, totalPrice, transactionId, trackingNumber);
    }
}

File: csharp-workshop/Part2.LightMocks/Domain/Order.cs
namespace Part2.LightMocks.Domain;

public record Order(
    string Id,
    string CustomerId,
    IReadOnlyList<OrderItem> Items,
    OrderStatus Status,
    double TotalPrice,
    string? TransactionId,
    string? TrackingNumber)
{
    public Order(
        string id,
        string customerId,
        IReadOnlyList<OrderItem> items,
        OrderStatus status,
        double totalPrice,
        string? transactionId,
        string? trackingNumber)
        : this(id, customerId, items.ToList().AsReadOnly(), status, totalPrice, transactionId, trackingNumber)
    {
    }

    public Order Confirm(double totalPrice) =>
        this with { Status = OrderStatus.Confirmed, TotalPrice = totalPrice };

    public Order Reject() =>
        this with { Status = OrderStatus.Rejected };
}
